torch/csrc/jit/codegen/cuda/arith.cpp

```diff
@@ -34,6 +34,37 @@ TensorView* maybe_broadcast_inner_to_rank(TensorView* t, size_t rank) {
   return t;
 }
 
+TensorView* maybe_broadcast_idxtv_inner_outter_to_rank(
```

Comment:
Why broadcasting is needed? I have a feeling that if we modify https://github.com/csarofeen/pytorch/blob/c9f8c1de5f3bd6ac0a0cce50569410abe2dfc209/torch/csrc/jit/codegen/cuda/root_domain_map.cpp#L65 to correctly handle pairwise root domain mapping, it might not be needed to broadcast here.

Reply:
yes, we could remove broadcast but we have to modify both PairwiseRootDomainMap::map and mapPointwiseOrReductionOp or if we keep broadcast we can reuse mapPointwiseOrReductionOp. Right now, current impl reuse mapPointwiseOrReductionOp and keep broadcast. Is it ok?

Reply:
@naoyam @csarofeen  What do you think? It looks like `index_select` is just a broadcast of gather:

```python
import torch

def my_index_select(self, dim, index):
    broadcast_shape = [1] * len(self.shape)
    broadcast_shape[dim] = index.shape[0]
    expand_shape = list(self.shape)
    expand_shape[dim] = index.shape[0]
    index = index.view(broadcast_shape).expand(expand_shape)
    return self.gather(dim, index)

x = torch.randn(5, 6, 7)
index = torch.tensor([1, 2, 3])

assert torch.eq(x.index_select(0, index), my_index_select(x, 0, index)).all()
assert torch.eq(x.index_select(1, index), my_index_select(x, 1, index)).all()
assert torch.eq(x.index_select(2, index), my_index_select(x, 2, index)).all()
```

And if we add broadcasting here, this PR effectively implements `gather` (which is good I guess). In the future gather PR, we might just rename the op name from `IndexSelectOp` to `GatherOp`.

Reply:
IIUC, `gather` is more flexible than `index_select`, but at this point since we assume inputs are fusion inputs, the added flexibility doesn't matter. Since we would eventually need to lift the assumption, don't we want to have them as separate operations?

Reply:
> IIUC, `gather` is more flexible than `index_select`, but at this point since we assume inputs are fusion inputs, the added flexibility doesn't matter. Since we would eventually need to lift the assumption, don't we want to have them as separate operations?

This is a good point. I agree.

Reply:
Xiang, I removed this broadcast in local branch(looks ok), but this issue https://github.com/csarofeen/pytorch/issues/2216 blocks me. Could we just keep the broadcast in this PR and then I will create another PR for index_select(removing broadcast) after issue2216 fixed?

Reply:
@naoyam Without broadcasting, we do have issues with transform replay. For example
```
T0[I0] ...
T1[I1, I2, I3] = input
T2[I1, I0, I3] = index_select(lookup_tv=T1, dim=1, index_tv=T0)
```
and if `T2` is replayed as `[(I1*I0*I3)/128, 128]`, what should `T0` be replayed as? Do we want `[I0/128, 128]`? Or a more general question is, do we want to enable the forwarding for all unmapped IterDomains in the consumer, instead of just broadcasting?

Reply:
I see. Looks like broadcasting the index tensor seems reasonable then. Alternatively, we could extend the transform forwarding for index_select, but I don't know which would be better. Broadcasting seems most straightforward for index_select, so I think that would make most sense.